src/main/java/emu/grasscutter/game/player/Player.java

```diff
@@ -390,6 +395,20 @@ public synchronized void setScene(Scene scene) {
         this.scene = scene;
     }
 
+    public void visitScene(int sceneId) {
+        if (visitedScenes.contains(sceneId)) return;
+        visitedScenes.add(sceneId);
+        val sceneTagData = GameData.getSceneTagDataMap().values();
+        val tagsToAdd = sceneTagData.stream()
+                .filter(tagData -> tagData.getSceneId() == sceneId && tagData.isDefaultValid())
+                .map(SceneTagData::getId)
+                .collect(Collectors.toSet());
+
+        val tagsToSet = this.sceneTags.getOrDefault(sceneId, new HashSet<>());
```

Comment:
instead of getOrDefault it might be better to use `.computeIfAbsent(partnerId, x -> new HashSet<>())`.
This will directly add the new element to the sceneTag Map, and return it, so no need to put it afterwards

Reply:
Got it! computeIfAbsent is a lot nicer than the read-modify-write template that I was going for